---
title: "SNP filtering - for stairwayplot"
output: github_document
---
keep samples from Weddell Sea, Amundsen Sea, Ross Sea, East Antarctica, Shetland Islands
<br>
here we start with the maxmissing05.recode.vcf generated in the previous SNP filtering pipeline (./turqueti/tarcap/SNPfiltering2/7_max_missing05) for demographic model to save some time  
<br>
define site ID, keep biallelic sites only and filter by minor allele count of 1:
```{bash, error=TRUE, results='hide'}
# first, define site ID
cd ./turqueti/tarcap/SNPfiltering3/7_max_missing05  

/sw/containers/bcftools-1.13.sif bcftools annotate --set-id '%CHROM\_%POS' maxmissing05.recode.vcf > maxmissing05_setID.vcf

# then, keep biallelic sites only
cd ./turqueti/tarcap/SNPfiltering3/8_biallelic_only  

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering3/7_max_missing05/maxmissing05_setID.vcf --min-alleles 2 --max-alleles 2 --out maxmissing05_setID_biallelic --recode --recode-INFO-all

# filter by minor allele count of 1
cd ./turqueti/tarcap/SNPfiltering3/9_mac01  

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering3/8_biallelic_only/maxmissing05_setID_biallelic.recode.vcf --mac 1 --out mac01.vcf --recode --recode-INFO-all
```
keep polymorphic sites with a maximum observed heterozygosity value of 0.7:
typical for RAD-loci data, for refs see https://onlinelibrary.wiley.com/doi/full/10.1111/j.1755-0998.2010.02967.x &
https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13314
```{r, eval=F, echo=T}
# calculate observed heterozygosity value per site
library(adegenet)
library(pegas)
library(vcfR)

# get a genind from the semi-filtered vcf file from bcftools

vcf2 <- read.vcfR("./mac01.vcf.recode.vcf")
oph_genind <- vcfR2genind(vcf2)
summary_genind = summary(oph_genind)

# get the Hobs per SNP 
summary_genind$Hobs 

write.csv(summary_genind$Hobs, file = "Hobs_per_SNP.csv")

# you can then sort the values in excel and save a list of the SNP (with site ID) with >0.7
```
```{bash, error=TRUE, results='hide'}
# Filter sites with maximum observed heterozygosity of 0.7
cd ./turqueti/tarcap/SNPfiltering3/10_max_obs_het07

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering3/9_mac01/mac01.vcf.recode.vcf --exclude site_obs_het_over_07.tsv --out max_het07.vcf --recode --recode-INFO-all
```
set ancestral alelles using outgroups in vcf file:  
see ./hpc/tarcap/set_ancestral_alleles.Rmd for further info
```{bash, error=TRUE, results='hide'}
# create an INFO file line for the new vcf file
echo '##INFO=<ID=AA,Number=1,Type=Character,Description="Ancestral allele">' > hdr.txt

# use bcftools to annotate the vcf file with the ancestral allele information
/sw/containers/bcftools-1.13.sif bcftools annotate -a ../out_aa.tab.gz \
 -c CHROM,POS,REF,ALT,INFO/AA -h hdr.txt -Oz \
 -o max_het07_aa.vcf.gz max_het07.vcf.recode.vcf.gz
```
check if setting AA has worked:
```{bash, error=TRUE, results='hide'}
/sw/containers/bcftools-1.13.sif query -f '%CHROM\t%POS\t%REF\t%ALT\t%INFO/AA\n' max_het07_aa.vcf.gz | less

# Count how many sites have ancestral allele information
/sw/containers/bcftools-1.13.sif view -e 'INFO/AA=="."' max_het07_aa.vcf.gz -H | wc -l   
```
remove Shetland Islands samples (depends on analyses):
```{bash, error=TRUE, results='hide'}
cd ./turqueti/tarcap/SNPfiltering3/11_rm_shetland

/sw/containers/vcftools-0.1.16.sif vcftools --gzvcf ./turqueti/tarcap/SNPfiltering3/10_max_obs_het07/max_het07_aa.vcf.gz --remove shetland_tur.tsv --out rmshetland_aa --recode --recode-INFO-all
```