---
title: "Stairway Plot"
output: github_document

## Prepare StairwayPlot input files using easySFS
https://github.com/isaacovercast/easySFS  
<br>
Run an easySFS trial run to see if we need to downproject (i.e. downsample) the SFS  
Whether or not you need to downproject your SFS depends on the amount of missing data you have. For more info on downprojection, see https://dadi.readthedocs.io/en/latest/user-guide/manipulating-spectra/  
<br>
Don't forget the number of samples is counted in haploids here, and we have a diploid species (P. turqueti)  
```{bash, error=TRUE, results='hide'}
cd ./turqueti/tarcap/analyses/stairwayplot/easysfs

./easySFS.py -i ./rmshetland_aa.vcf -p ./easysfs_popmap.txt -a --unfolded --preview

#the existing P. turqueti dataset has very few missing data so we have already maximised the number of segregating sites. We don't need to downproject our SFSs here. This will be different across different projects
```

Generate SFSs:  
<br>
Set 'proj' based on the number of haploids we have per pop  
<br>
Sample size per pop (in diploids):  
Ross Sea; n=10  
East Antarctica; n=5  
Amundsen Sea; n=4  
Weddell Sea; n=18  

```{bash, error=TRUE, results='hide', eval=FALSE}
cd ./turqueti/tarcap/analyses/stairwayplot/easysfs

./easySFS.py -i ./rmshetland_aa.vcf -p ./easysfs_popmap.txt -a --unfolded --proj=20,10,8,36  -o ./turqueti_stairwayplot --prefix RS_EA_AM_WS -f -v

#easySFS generates SFSs in dadi and fastsimcoal formats
#for stairwayplot, we stick with the dadi format
#easySFS also generates 1D-SFS (1 SFS per pop), 2D-SFS (pairwise SFS between every 2 pops), and 4D-SFS (a single SFS containing all the information of the 4 pops we entered here)
#For stairwayplot, we use 1D-SFS
```

Visualise the SFSs for sanity checks:  
<br>
Here I am viewing the 2D-SFSs but you can easily visulaise 1D-SFS too  
I am using a function within dadi for visualisation (https://dadi.readthedocs.io/en/latest/), you can use a simple ggplot in R for this as well

```{bash, error=TRUE, results='hide', eval=FALSE}
ipython

import dadi
import pylab

#e.g. view 2D-SFS between Amundsen Sea and Weddell Sea pops
fs = dadi.Spectrum.from_file("./turqueti/tarcap/analyses/stairwayplot/easysfs/dadi/Amundsen-Weddell.sfs")

dadi.Plotting.plot_single_2d_sfs(fs, vmin = 0.05)

pylab.show()
```

## Preprating stairwayplot input files
First we need to figure out L (the length of genome we have kept after SNP filtering, including polymorhpic and monomorphic sites)
```{bash, error=TRUE, results='hide'}
## Keep 1 SNP per locus to see how many loci we have kept after SNP filtering
cd ./turqueti/tarcap/SNPfiltering3/11_rm_shetland

/sw/containers/vcftools-0.1.16.sif --vcf rmshetland_aa.vcf --thin 1000  --out thin1000.vcf --recode --recode-INFO-all
/sw/containers/vcftools-0.1.16.sif --vcf thin1000.vcf.recode.vcf --missing-site (n of loci kept = 5780)

##grab the ID of the retained loci (first column of out.lmiss)
awk '{ print $1 }' out.lmiss > list_locus_temp.txt

#remove the header line of list_locus_temp.txt
echo "$(tail -n +2 list_locus_temp.txt)" > list_locus.txt

##Calculate the number of bases in these 5780 loci from unique_consensus_loci.fa.fai

#first, grab these 5780 loci from unique_consensus_loci.fa.fai
awk ' FILENAME=="unique_consensus_loci.fa.fai" {arr[$1]=$0; next}
        FILENAME=="list_locus.txt"  {print arr[$1]} ' unique_consensus_loci.fa.fai list_locus.txt > keep_loc_file.fai
        
#get L from sum of column 2 of keep_loc_file.fai
awk '{s+=$2}END{print s}' keep_loc_file.fai #L=2058671
```

Generate .blueprint file per population (input of stairwayplot)  
<br>
see ./hpc/stairwayplot for other .blueprint files

## Run stairwayplot

```{bash, error=TRUE, results='hide', eval=FALSE}
module load java

java -cp ./stairway_run/stairway_plot_es Stairbuilder tur_ross_unfolded.blueprint #this generates tur_ross_unfolded.blueprint.sh

bash tur_ross_unfolded.blueprint.sh 
```

Then plot results using ggplot2 for better visualisation